### PR TITLE
KT-22063 Add intrinsics for javaObjectType and javaPrimitiveType

### PR DESCRIPTION
**Pull request by mikaelpeltier**
- Add new intrinsics into both backends

Fix of https://youtrack.jetbrains.com/issue/KT-22063

### CODE REVIEW

File: compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+class KClassJavaPrimitiveTypeProperty : IntrinsicPropertyGetter() {
+    override fun generate(resolvedCall: ResolvedCall<*>?, codegen: ExpressionCodegen, returnType: Type, receiver: StackValue): StackValue? {
+        val receiverValue = resolvedCall!!.extensionReceiver as? ExpressionReceiver ?: return null
+        val classLiteralExpression = receiverValue.expression as? KtClassLiteralExpression ?: return null
+        val receiverExpression = classLiteralExpression.receiverExpression ?: return null
+        val lhs = codegen.bindingContext.get(DOUBLE_COLON_LHS, receiverExpression) ?: return null
+        val lhsType = codegen.asmType(lhs.type)
+        if (TypeUtils.isTypeParameter(lhs.type)) {
+            // Reified mechanism is not yet able to generate code that should generate a null value or a field access to TYPE based
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: I'd rephrase this comment to simply say something like `// TODO: add new operation kind to ReifiedTypeInliner.OperationKind`

File: compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+                val receiverType = receiverStackValue.type
+                when {
+                    receiverType == Type.VOID_TYPE -> {
+                        receiverStackValue.put(Type.VOID_TYPE, iv)
+                        iv.aconst(null)
+                    }
+                    AsmUtil.isPrimitive(receiverType) -> {
+                        receiverStackValue.put(receiverType, iv)
+                        AsmUtil.pop(iv, receiverType)
+                        iv.getstatic(AsmUtil.boxType(receiverType).getInternalName(), "TYPE", "Ljava/lang/Class;");
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: Use `internalName` instead of `getInternalName()` here and elsewhere in this file; also drop the `;`

File: compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+                            iv.getstatic(receiverType.getInternalName(), "TYPE", "Ljava/lang/Class;");
+                        } else {
+                            iv.aconst(null)
+                        }
+                    }
+                }
+            } else {
+                if (AsmUtil.isPrimitive(lhsType)
+                    || AsmUtil.unboxPrimitiveTypeOrNull(lhsType) != null
+                    || AsmTypes.VOID_TYPE.equals(lhsType)
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: `==` instead of `equals`

File: compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+            return null
+        }
+
+        return StackValue.operation(returnType) { iv ->
+            if (lhs is DoubleColonLHS.Expression && !lhs.isObjectQualifier) {
+                val receiverStackValue = codegen.gen(receiverExpression)
+                val receiverType = receiverStackValue.type
+                when {
+                    receiverType == Type.VOID_TYPE -> {
+                        receiverStackValue.put(Type.VOID_TYPE, iv)
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: Can be simplified similarly to the `when` in `KClassJavaObjectTypeProperty`

File: compiler/frontend.java/src/org/jetbrains/kotlin/resolve/jvm/AsmTypes.java [hunk 5]
```diff
@@ -34,6 +34,7 @@
     public static final Type NUMBER_TYPE = getType(Number.class);
     public static final Type BOOLEAN_WRAPPER_TYPE = getType(Boolean.class);
     public static final Type CHARACTER_WRAPPER_TYPE = getType(Character.class);
+    public static final Type VOID_TYPE = getType(Void.class);
```
Comment (udalov) [inline compiler/frontend.java/src/org/jetbrains/kotlin/resolve/jvm/AsmTypes.java]: There's already `Type.VOID_TYPE`, please use it instead

File: compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                        receiverStackValue.put(receiverType, iv)
+                        AsmUtil.pop(iv, receiverType)
+                        if (AsmUtil.unboxPrimitiveTypeOrNull(receiverType) != null) {
+                            iv.getstatic(receiverType.getInternalName(), "TYPE", "Ljava/lang/Class;");
+                        } else {
+                            iv.aconst(null)
+                        }
+                    }
+                }
+            } else {
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: I suppose you can merge this to `} else if { ...`

File: compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+                if (AsmUtil.isPrimitive(lhsType)
+                    || AsmUtil.unboxPrimitiveTypeOrNull(lhsType) != null
+                    || AsmTypes.VOID_TYPE.equals(lhsType)
+                ) {
+                    var type = lhsType
+                    if (AsmUtil.isPrimitive(lhsType)) {
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: This `if` is unnecessary, you can call `boxType` even on an already boxed type

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/KClassJavaObjectTypeProperty.kt [hunk 8]
```diff
@@ -0,0 +1,66 @@
+/*
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/KClassJavaObjectTypeProperty.kt]: Please fix inspections and reformat this file and `KClassJavaPrimitiveTypeProperty.kt`

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 9]
```diff
@@ -0,0 +1,85 @@
...
+                            if (AsmUtil.unboxPrimitiveTypeOrNull(extensionReceiverType) != null) {
+                                v.getstatic(extensionReceiverType.getInternalName(), "TYPE", "Ljava/lang/Class;");
+                            } else {
+                                v.aconst(null)
+                            }
+                        }
+                    }
+                } else if (extensionReceiver is IrClassReference) {
+                    assert(extensionReceiver is IrClassReference)
+                    val classType = (extensionReceiver as IrClassReference).classType
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: After three checks, I'd be surprised if it wasn't a `IrClassReference`... :) This cast and the assert above are not necessary

File: compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+                        codegen.gen(extensionReceiver!!, data)
+                        val mapToCallableMethod =
+                            context.state.typeMapper.mapToCallableMethod(expression.descriptor as FunctionDescriptor, false)
+                        mapToCallableMethod.genInvokeInstruction(codegen.mv)
+                    } else {
+                        var lhsType = codegen.typeMapper.mapType(classType)
+                        if (AsmUtil.isPrimitive(lhsType)
+                            || AsmUtil.unboxPrimitiveTypeOrNull(lhsType) != null
+                            || AsmTypes.VOID_TYPE.equals(lhsType)) {
+                            if (AsmUtil.isPrimitive(lhsType)) {
```
Comment (udalov) [inline compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: Similarly, this `if` is not needed

**Comment by udalov** [comment]
I've reviewed the main codegen part and the tests, and everything seems OK.

@max-kammerer Please review the `compiler/backend.jvm` part. I'm not sure about the fact that the code is copy-pasted and whether we can do something about it, but seems like all other new backend intrinsics are copy-pasted as well...

File: compiler/testData/codegen/box/intrinsics/kt22063_1.kt [hunk 11]
```diff
@@ -0,0 +1,85 @@
+// IGNORE_BACKEND: JS
```
[the remark this answers is not in the document]
Comment (mikaelpeltier) [reply compiler/testData/codegen/box/intrinsics/kt22063_1.kt]: Done

File: compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt [hunk 12]
```diff
@@ -23,51 +23,50 @@ class KClassJavaPrimitiveTypeProperty : IntrinsicPropertyGetter() {
...
                     else -> {
-                        receiverStackValue.put(receiverType, iv)
-                        AsmUtil.pop(iv, receiverType)
                         if (AsmUtil.unboxPrimitiveTypeOrNull(receiverType) != null) {
-                            iv.getstatic(receiverType.getInternalName(), "TYPE", "Ljava/lang/Class;");
+                            AsmUtil.pop(iv, receiverType)
+                            iv.getstatic(receiverType.internalName, "TYPE", "Ljava/lang/Class;")
                         } else {
-                            iv.aconst(null)
+                            if (receiverType == AsmTypes.OBJECT_TYPE || receiverType == AsmTypes.NUMBER_TYPE) {
```
Comment (udalov) [inline compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: Instead of this, can we check the receiver expression type _before_ applying the intrinsic and return null if it's not primitive? E.g.
```
if (lhs is DoubleColonLHS.Expression && !lhs.isObjectQualifier) {
    val receiverType = codegen.bindingContext.getType(receiverExpression) ?: return null
    if (!KotlinBuiltIns.isPrimitiveTypeOrNullablePrimitiveType(receiverType)) return null
}
```

Comment (mikaelpeltier) [reply compiler/backend/src/org/jetbrains/kotlin/codegen/intrinsics/KClassJavaPrimitiveTypeProperty.kt]: Done.

**Comment by udalov** [comment]
Thanks! Merged manually.
